zboxcore/sdk/registry_file_manager.go

```diff
@@ -0,0 +1,203 @@
+
+	err = os.WriteFile(tempLocal, data, 0600)
+	if err != nil {
+		return errors.New("update_registry_file_failed", "Failed to create file locally for upload: "+err.Error())
+	}
+
+	err = a.fileStorer.StartChunkedUpload(os.TempDir(), tempLocal, a.registryFilePath, nil, isUpdate, false, "", false)
+	if err != nil {
+		return errors.New("update_registry_file_failed", "Failed to upload registry file: "+err.Error())
+	}
+
+	return nil
+}
+
+// Get retrieves the registry file contents and its last updated timestamp.
+func (a *registryFileStore) Get() ([]byte, common.Timestamp, error) {
+	rfLastUpdateTime, err := a.GetLastUpdateTimestamp()
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to check existence of registry file: "+err.Error())
+	}
+
+	// no registry file yet
+	// return empty
+	if rfLastUpdateTime == common.Timestamp(0) {
+		return []byte{}, rfLastUpdateTime, nil
+	}
+
+	// download registry file
+	tempLocal := filepath.Join(os.TempDir(), fmt.Sprintf("starred_dl_%d", time.Now().UnixNano()))
+	cb := newRegistryFileCB()
+
+	err = a.fileStorer.DownloadFile(tempLocal, a.registryFilePath, cb)
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to start download of registry file: "+err.Error())
+	}
+
+	<-cb.Wait() // wait for download to complete
+	if cb.err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to download registry file: "+cb.err.Error())
+	}
+
+	f, err := os.Open(tempLocal)
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to open downloaded registry file: "+err.Error())
+	}
+
+	defer os.Remove(tempLocal)
+	defer f.Close()
+
+	bt, err := io.ReadAll(f)
+	if err != nil {
+		return nil, common.Timestamp(0), errors.New("get_registry_file_failed", "Failed to read downloaded registry file: "+err.Error())
+	}
+
+	return bt, rfLastUpdateTime, nil
+}
+
+// GetLastUpdateTimestamp retrieves the last update timestamp of the registry file.
+// Returns 0 as timestamp when no registry file exist.
+func (a *registryFileStore) GetLastUpdateTimestamp() (common.Timestamp, error) {
+	rf, err := a.getRegistryFile()
+	if err != nil {
+		return common.Timestamp(0), errors.New("get_last_update_timestamp_failed", "Failed to check existence of registry file: "+err.Error())
+	}
+
+	// when no registry yet, return beginning of epoch as default last update timestamp.
+	if rf == nil {
+		return common.Timestamp(0), nil
+	}
+
+	lastUpdate, err := time.Parse(time.RFC3339, rf.UpdatedAt)
+	if err != nil {
+		return common.Timestamp(0), errors.New("get_last_update_timestamp_failed", "Failed to parse last updated timestamp of registry file: "+err.Error())
+	}
+
+	return common.Timestamp(lastUpdate.Unix()), nil
+}
+
+func (a *registryFileStore) getRegistryFile() (*ListResult, error) {
+	dir, err := a.fileStorer.ListDir(a.registryFilePath)
```

Comment:
This ListDir will only return one file (the registry file) right? The method name is a little confusing, I would thought it would list all the files in a dir. Perhaps rename it to List and wraps it to call the ListDir to return the file.

Reply:
`fileStorer.ListDir()` here is referring to the `Allocation.ListDir()` function

https://github.com/0chain/gosdk/blob/d7645bf6f88b1e76ac06e59dccb37a81a1af3143/zboxcore/sdk/allocation.go#L654

I'll just rename `getRegistryFile() (*ListResult, error)` to be `getRegistryFileLastUpdate() (common.Timestamp, error)`. This way it does need to leak `*ListResult`. Only important for users of that method is to know when was the last file update.